quickwit-storage/src/split.rs

```diff
@@ -0,0 +1,469 @@
+// Copyright (C) 2021 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::io::{self, ErrorKind, SeekFrom};
+use std::ops::Range;
+use std::path::{Path, PathBuf};
+
+use async_trait::async_trait;
+use futures::{stream, StreamExt};
+use rusoto_core::ByteStream;
+use tokio::io::{AsyncReadExt, AsyncSeekExt};
+use tokio_util::io::ReaderStream;
+
+use crate::{BundleStorageFileOffsets, PutPayload};
+
+/// Payload of a split which builds the split bundle and hotcache on the fly and streams it to the
+/// storage.
+#[derive(Clone)]
+pub struct SplitPayload {
+    payloads: Vec<Box<dyn PutPayload>>,
+    /// bytes range of the footer (hotcache + bundle metadata)
+    pub footer_range: Range<u64>,
+}
+
+async fn range_byte_stream_from_payloads(
+    payloads: &[Box<dyn PutPayload>],
+    range: Range<u64>,
+) -> io::Result<ByteStream> {
+    let mut bytestreams: Vec<_> = Vec::new();
+
+    let payloads_and_ranges =
+        chunk_payload_ranges(payloads, range.start as usize..range.end as usize);
+
+    for (payload, range) in payloads_and_ranges {
+        bytestreams.push(
+            payload
+                .range_byte_stream(range.start as u64..range.end as u64)
+                .await?,
+        );
+    }
+
+    let concat_stream = ByteStream::new(stream::iter(bytestreams).flatten());
+    Ok(concat_stream)
+}
+
+#[async_trait]
+impl PutPayload for SplitPayload {
+    fn len(&self) -> u64 {
+        self.payloads.iter().map(|payload| payload.len()).sum()
+    }
+
+    async fn range_byte_stream(&self, range: Range<u64>) -> io::Result<ByteStream> {
+        range_byte_stream_from_payloads(&self.payloads, range).await
+
+        // let requested_payload_ranges = chunk_range(&absolute_payload_ranges, range.start as usize
```

Comment:
Remove commented code.

Reply:
nitpick: this is the unique call of `range_byte_stream_from_payloads` so you can directly put the implementation in the `SplitPayload.